devicetree: use stable identifiers for LLEXT-exported DT object names

pillo79 opened the pull request:
This is the candidate implementation for RFC #83800.

pillo79 commented:
Compliance check found a leftover `continue` introduced by commit d25e5c20a07203443b2e9fafe30bf8eef852ed23, 3 years ago. 
It seems not many people dare to venture here! :slightly_smiling_face:

pillo79 commented:
**v2**:
- rebased to remove commits merged in #78508 from this PR
- compliance issue got fixed a few days ago in 371a48ac16f

mathieuchopstm commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -90,6 +91,12 @@
 from devicetree.grutils import Graph
 from devicetree._private import _slice_helper
 
+def _compute_hash(path: str) -> str:
+    # Calculates the hash associated with the node's full path.
+    hasher = hashlib.sha256()
+    hasher.update(path.encode())
+    return hasher.hexdigest()
+
```
A few thoughts:
* we could save space (~33%) by base64-encoding the hash's raw bytes
  * Limitation: the resulting hash must be a valid C identifier
    * base64 uses `+`, `/` and `=` which are illegal...
    * base32 saves only ~20% but doesn't have this problem
* appending a known-unique value (e.g., node address) to the hash string would guarantee that collisions can *never* happen
  * not sure this is worth the trouble with how unlikely SHA-256 collisions are, though

pillo79 replied:
Thanks for the idea!

> base64 uses `+`, `/` and `=` which are illegal...

True, but we don't really need an invertible transformation, just a hash, so we can cheat a little:  
 * `=` is only used in padding, so we can drop it. 
 * Replacing both `+` and `/` with the legal symbol `_` would make clashes _a little_ more probable, but still vanishingly small: ChatGPT said with a convincing formula that it's 1.23 times more probable than a direct SHA256 collision (5.5E-78...).

At the same time, though, the identifier would become "only" 43 chars. Looks like a nice tradeoff. :tada:

mathieuchopstm commented on include/zephyr/llext/symbol.h:
```diff
@@ -113,32 +125,45 @@ struct llext_symtable {
-		.name = STRINGIFY(x), .addr = (const void *)&x,			\
+/* LLEXT-enabled application: export symbols */
+#define Z_EXPORT_SYMBOL_NAMED(sym_ident, sym_name)					\
+	static const STRUCT_SECTION_ITERABLE(llext_const_symbol, sym_ident ## _sym) = {	\
```
I see why it's been done this way (`sym_name` is a string), but wouldn't using `sym_ident` to form the `llext_const_symbol`'s name prevent exporting the same symbol twice under different names? (at least if both are in the same translation unit)

I think taking the `sym_name` in token form and `STRINGIFY()`ing it in the macro would be acceptable, if that can help solve this issue.

Same remark for SLID version of the macro.

pillo79 replied:
Why would you have to export the same symbol twice with two names? One should be enough for everybody! :hand_over_mouth: 

... seriously, this PR wanted to transparently rename symbols when appropriate, so I wasn't thinking about the use case you mentioned (there's always one `EXPORT_SYMBOL` for each identifier). 
It makes sense though, I will have to think about using the `sym_name` - I remember there was a problem that made me use strings and not tokens, but I'm not sure what that was.

mathieuchopstm reviewed (commented):
Just stumbled across this PR, and I must say it's a neat idea!
A few thoughts:

pillo79 reviewed (commented):
Thanks for the feedback @mathieuchopstm . Had so many open PRs these days I lost track of this one! :man_facepalming: I need to finish it and push it on.

pillo79 committed:
llext: add custom name variant to symbol export macros

Add a new set of macros that allow customizing the symbol name when
exporting symbols. This is useful when the symbol name that extensions
need to look up is different from the identifier used in the base image.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
device: add CONFIG_LLEXT_EXPORT_DEV_IDS_BY_HASH option

This new option allows to export devices using identifiers generated
from the hash of the devicetree node path, instead of the device's
ordinal number. Identifiers generated this way are stable across
rebuilds.

Add new test cases to test this new option.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
edtlib: add "hash" attribute to nodes

Add a new "hash" attribute to all Devicetree EDT nodes. The hash is
calculated on the full path of the node; this means that its value
remains stable across rebuilds.
The hash is checked for uniqueness among nodes in the same EDT.

This computed token is then added to `devicetree_generated.h` and made
accessible to Zephyr code via a new DT_NODE_HASH(node_id) macro.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
**v3**, almost ready for public review:
* rebased to current Zephyr
* applied both suggestions from @mathieuchopstm:
  * switched from plain hex to base64 encoding to shorten the generated identifiers
  * by naming the llext symbols according to the desired alias, the same identifier can now have any number of export aliases

rruuaanng commented on include/zephyr/devicetree.h:
```diff
@@ -242,6 +242,16 @@
+ * @param node_id node identifier
+ * @return hash value as a preprocessor token
+ */
+#define DT_HASH(node_id) DT_CAT(node_id, _HASH)
```
It's repetitive, it appears in #83748.

rruuaanng commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -90,6 +92,14 @@
 from devicetree.grutils import Graph
 from devicetree._private import _slice_helper
 
+def _compute_hash(path: str) -> str:
```
It's repetitive, it appears in #83748.

rruuaanng commented on scripts/dts/python-devicetree/tests/test_edtlib.py:
```diff
@@ -241,6 +241,16 @@ def test_hierarchy():
 
     assert edt.get_node("/parent/child-1").children == {}
 
+def test_hashes():
```
It's repetitive, it appears in #83748.

rruuaanng commented:
It seems that #83748 is the main work. Can we close this PR and create an issue for the RFC?

pillo79 commented:
> It seems that https://github.com/zephyrproject-rtos/zephyr/pull/83748 is the main work.

In fact it's the opposite - this is the main work and #83748 is only the first commit of this series. 
I wanted to split the EDT and C/LLEXT changes to make it easier to review and discuss upon, but alas, ended up creating more confusion! :man_facepalming:

I have now submitted the RFC as issue #83800.

pillo79 commented:
**v4**, incorporating the feedback received in #83748:
* renamed the `DT_HASH` macro to `DT_NODE_HASH`
* used a more direct solution to obtain identifier-compatible base64 encodings 
* moved the hash test suite from inside edtlib to the Twister-based `tests/lib/devicetree`

rruuaanng commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -338,6 +338,16 @@ ZTEST(devicetree_api, test_has_alias)
 
+ZTEST(devicetree_api, test_node_hashes)
+{
+	zassert_str_equal(TO_STRING(DT_NODE_HASH(DT_ROOT)),
```
Based on the content of `gen_defines`, it seems that the generated macro should be `#define DT_... "il7asoJjJEMhngUeSt4tHVu8Zxx4EFG_FDeJfL3_oPE"`. Why is it necessary to convert the symbol into a string in this case?

pillo79 replied:
The `gen_defines.py` in fact exports hashes as tokens, not strings. Here are a few sample ones I extracted from a `devicetree_generated.h` with this PR:
```
...
#define DT_N_S_soc_S_gpio_40013000_HASH 7HLGxCJFZAxHOOrLys2k53iNEtZBvFanrB0PRFnjT3w
#define DT_N_S_soc_S_gpio_40012000_HASH 7jBOD_T2CWBDoxYU2ETQTv5jO4fSyafrS9t1xW6F5gA
#define DT_N_S_soc_S_gpio_40011000_HASH eDfnG7tCZbQPQkasUZ0ThNugDf4R0ZHWfFavc0Bu4c4
#define DT_N_S_soc_S_gpio_40010000_HASH PxXZfhmPKDGCr74hZk7An6sUAWH9x1Oj9oGnAGdkti8
...
```

The reason for exporting tokens is to be able to use these in identifiers as well as strings. These are passed as part of the `name` parameter in  the new `EXPORT_SYMBOL_NAMED(ident, name)` macro and end up inside the name of the constants autogenerated in `EXPORT_SYMBOL` to hold the LLEXT-related information:
```
...
00008f50 l     O llext_const_symbol_area        00000008 __llext_sym___device_dts_7HLGxCJFZAxHOOrLys2k53iNEtZBvFanrB0PRFnjT3w
00008f60 l     O llext_const_symbol_area        00000008 __llext_sym___device_dts_7jBOD_T2CWBDoxYU2ETQTv5jO4fSyafrS9t1xW6F5gA
00008f88 l     O llext_const_symbol_area        00000008 __llext_sym___device_dts_eDfnG7tCZbQPQkasUZ0ThNugDf4R0ZHWfFavc0Bu4c4
00008f70 l     O llext_const_symbol_area        00000008 __llext_sym___device_dts_PxXZfhmPKDGCr74hZk7An6sUAWH9x1Oj9oGnAGdkti8
...
```

Additionally, this choice allows using the same `EXPORT_SYMBOL_NAMED(ident, name)` macro, outside of DT devices, to export the same `ident` as multiple `name`s if that is ever required.

pillo79 commented:
No further feedback received on the RFC or this PR so marking this as open for full review.

rruuaanng reviewed (approved):
Apologies for my limited familiarity with llext. So, I will leave this task to others. Thank you for your work on this.
